### PR TITLE
Add multi class model and Platt scaling

### PR DESCRIPTION
**Pull request by bytesnake**
Composing model for binary to multi-class transformation, WIP

 * [x] finish `MultiClassModel`
 * [x] add [Platt scaling](https://en.wikipedia.org/wiki/Platt_scaling) for proper SVM's probability values
 * [x] add example
 * [x] improve documentation

# Example

```rust
// we have to specify that we want to predict probabilities
// `Svm::<_, bool>` would also be possible and avoid Platt scaling.
let params = Svm::<_, Pr>::params()
  .gaussian_kernel(30.0);

let model = train.one_vs_all()?
    .into_iter()
    .map(|(l, x)| (l, params.fit(&x).unwrap()))
    .collect::<MultiClassModel<_, _>>();

// predict with validation dataset, the prediction has type `usize`
let pred = model.predict(&valid);

// create and print a confusion matrix
let cm = pred.confusion_matrix(&train)?;
println!("{:?}", cm);
```

### CODE REVIEW

**Review by Sauro98 — APPROVED** [review]
Good work, nice to have a unified multi-class model 🚀 

I would just add a note about the difference in the SVM params with `bool` or `Pr` also in the parameters' documentation so that the user doesn't have to go look in the example.

File: algorithms/linfa-svm/src/classification.rs [hunk 1]
```diff
@@ -260,45 +314,76 @@ impl_oneclass!(Array2<F>, CountedTargets<(), ArrayView2<'a, ()>>);
...
 }
 
+/// Predict a probability with a feature vector
+impl<'a, F: Float> Predict<ArrayView1<'a, F>, bool> for Svm<F, bool> {
+    fn predict(&self, data: ArrayView1<'a, F>) -> bool {
+        let val = self.weighted_sum(&data) - self.rho;
+
+        val >= F::zero()
+    }
+}
```
[the remark this answers is not in the document]
Comment (bytesnake) [reply algorithms/linfa-svm/src/classification.rs:349]: yes

File: algorithms/linfa-svm/examples/winequality_multi.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    let params = Svm::<_, Pr>::params()
+        //.pos_neg_weights(5000., 500.)
+        .gaussian_kernel(30.0);
+
+    let model = train
+        .one_vs_all()?
+        .into_iter()
+        .map(|(l, x)| (l, params.fit(&x).unwrap()))
+        .collect::<MultiClassModel<_, _>>();
```
[the remark this answers is not in the document]
Comment (bytesnake) [reply algorithms/linfa-svm/examples/winequality_multi.rs:21]: would be definitely possible, perhaps add it later?

[the remark this answers is not in the document]
Comment (Sauro98) [reply algorithms/linfa-svm/examples/winequality_multi.rs:21]: 👍🏻

**Comment by quietlychris** [comment]
I saw that you linked the paper at https://www.csie.ntu.edu.tw/~cjlin/papers/plattprob.pdf in your explanation. It's totally possible that I missed it, but I couldn't find the first half of the 
```rust
    // avoid numerical problems for large f_apb
    if f_apb >= 0.0 {
        Pr((-f_apb).exp() / (1.0 + (-f_apb).exp()))
    } else {
        Pr(1.0 / (1.0 + f_apb.exp()))
    }
```
loop that you use for numerical stability with `f_apb` in the numerator in either that article or on the Wikipedia link. From first principles, it makes sense that it would work, but just wasn't sure if that was something you found in another paper that you might want to link as well. It's probably not necessary either way, but just got me curious. Otherwise, everything else looks great :)

**Review by quietlychris — APPROVED** [review]

**Comment by bytesnake** [comment]
> Good work, nice to have a unified multi-class model rocket
> 
> I would just add a note about the difference in the SVM params with `bool` or `Pr` also in the parameters' documentation so that the user doesn't have to go look in the example.

good point :+1: I'm not 100% sure why the compiler can't figure out when to use `bool` vs `Pr` though

> I saw that you linked the paper at https://www.csie.ntu.edu.tw/~cjlin/papers/plattprob.pdf in your explanation. It's totally possible that I missed it, but I couldn't find the first half of the
> 
> ```rust
>     // avoid numerical problems for large f_apb
>     if f_apb >= 0.0 {
>         Pr((-f_apb).exp() / (1.0 + (-f_apb).exp()))
>     } else {
>         Pr(1.0 / (1.0 + f_apb.exp()))
>     }
> ```
> 
> loop that you use for numerical stability with `f_apb` in the numerator in either that article or on the Wikipedia link. From first principles, it makes sense that it would work, but just wasn't sure if that was something you found in another paper that you might want to link as well. It's probably not necessary either way, but just got me curious. Otherwise, everything else looks great :)

yes you just have to multiply both sides with `(-f_apb).exp()`. I don't have a source right now, but probably similar to stable sigmoid implementations. Will find something :)

thank you both, I will complete the PR now and then merge

**Pull request merged**